consensus/raft/raft.go

```diff
@@ -427,6 +448,147 @@ func (rw *raftWrapper) Peers() ([]string, error) {
+	// For now assuming 3 substrings means snap file
+	splitA := strings.Split(pathA, "-")
+	splitB := strings.Split(pathB, "-")
+
+	if len(splitA) == 3 && len(splitB) != 3 {
+                // Raftsnapshots come first
+		return true
+	}
+
+	if len(splitB) == 3 && len(splitA) != 3 {
+		return false
+	}
+	
+	if len(splitB) != 3 && len(splitA) != 3 {
+		return pathA < pathB
+	}
+
+	// Both are raft snapshot files 
+	// The most recent snapshot belongs at the front of the slice
+	return splitA[2] > splitB[2]
+}
+
+//ExistingStateReader does a best effort search for existing raft state files
+//and returns the bytes of the latest raft snapshot
+func ExistingStateReader(cfg *Config) (io.Reader, error){
+	snapPath := "snapshots"
+	stateFilePath := "state.bin"
+
+	if cfg.BaseDir == "" || cfg.DataFolder == "" {
+		return nil, errors.New("Data location not specified in cfg")
+	}
+	snapShotDir := filepath.Join(cfg.BaseDir, cfg.DataFolder, snapPath)
+	
+	// take latest chronological file in the snapshot directory
+	snapFiles, err := ioutil.ReadDir(snapShotDir)
+	if len(snapFiles) == 0 {
+		return nil, errors.New("No snapshots saved")
+	}
+	sort.Sort(byRaftSnapFilename(snapFiles))
+	splitName := strings.Split(snapFiles[0].Name(), "-")
+	if len(splitName) != 3 {
+		return nil, errors.New("No snapshots saved")
+	}
+
+	// look for state file in latest snapshot
+	snapShotStateFile := filepath.Join(snapShotDir, snapFiles[0].Name(), stateFilePath)
+	rawBytes, err := ioutil.ReadFile(snapShotStateFile)
+	if err != nil {
+		return nil, err
+	}
+
+	var state interface {}
+	err = decodeState(rawBytes, &state)
+	if err != nil {
+		return nil, err
+	}
+	stateBytes, err := json.Marshal(state)
+	if err != nil {
+		return nil, err
+	}
+	r := bytes.NewReader(stateBytes)
+	return r, nil
+}
+
+//Reset saves a raft snapshot containing newState to be loaded on restart.
+//Only call when Raft is shutdown.
+func Reset(newState state.State, cfg *Config, raftDataPath string, peers []peer.ID) error{
+	err := cleanupRaft(raftDataPath)
+	if err != nil {
+		return err
+	}
+	snapshotStore, err := hraft.NewFileSnapshotStoreWithLogger(raftDataPath, RaftMaxSnapshots, nil)
+	if err != nil {
+		return err
+	}
+
+	serverAddr := hraft.ServerAddress(peer.IDB58Encode(peers[len(peers) -1]))
+	_, dummyTransport := hraft.NewInmemTransport(serverAddr)
+	var raftSnapVersion hraft.SnapshotVersion
+	raftSnapVersion = 1         // As of v1.0.0 this is always 1                                     
+	raftIndex       := uint64(1) // We reset history to the beginning                                                    
+	raftTerm        := uint64(1) // We reset history to the beginning
+	configIndex     := uint64(1) // We reset history to the beginning
+	srvCfg := MakeServerConf(peers)
+	sink, err := snapshotStore.Create(raftSnapVersion, raftIndex, raftTerm, srvCfg, configIndex, dummyTransport)
+	if err != nil {
+		return err
+	}
+	newStateBytes, err := encodeState(newState)
```

Comment:
let this function receive an already encoded state, perhaps a https://godoc.org/github.com/hashicorp/raft#FSMSnapshot object, which is what the FSM.Snapshot() provides.

Also probably rename to `SnapshotReset`. `Reset` sounds like a thing which could be maybe used by `Consensus` for something less specific in the future.

Reply:
@hsanjuan Hmmm correct me if I'm wrong but if this function receives an encoded state that implies the caller (ipfs-cluster-service/main.go) will be creating a dummy libp2p-raft object, loading up its FSM with state and then calling Snapshot().  This is actually what I started with before refactoring after your first round of comments, and I think you were right then that that approach forces the main function to be too raft aware.  

I do get that we would rather keep the encode / decode functionality confined to libp2p raft.  I was thinking maybe we could export Encode and Decode?  Writing the first draft of this showed me that the current libp2p-raft interface makes you jump through hoops to get encodings of states but at the end of the day leaks them out anyway so we wouldn't be changing what users can do, just how easy they can do it.  

Let me know what you think or if I am missing something here.

Reply:
Agree with the name change 👍

Reply:
Hmm, fair enough, you've convinced me. Let's export Encode and Decode!

You will need to make the change in libp2p-raft, do `gx publish -f` there, then take the hash and `gx update go-libp2p-raft <newhash>` here. Once everything is good we'll merge there, make a release, import a new final hash here. Let me know if you have problems with it.